### PR TITLE
Data utilities from OpteraLib

### PR DESCRIPTION
**Pull request by 0ptera**
Data stage functions from OpteraLib.

Please review and squash merge when done.

### CODE REVIEW

**Comment by raiguard** [comment]
I thought we decided not to pollute the global namespace? We shouldn't encourage people to create globals willy-nilly. I think you should have to require the file locally whenever you want to use it, just like in the control stage.

**Comment by 0ptera** [comment]
That's why I put it up for review.

I couldn't remember how that trick worked to enforce unique local registration in data.

File: data/utilities.lua [hunk 1]
```diff
@@ -0,0 +1,119 @@
+
+---@class PrototypeBase https://wiki.factorio.com/PrototypeBase
+---@class IconSpecification https://wiki.factorio.com/Types/IconSpecification
+
```
Comment (raiguard) [inline data/utilities.lua:5]: Please follow the [style guide's bit on modules](https://github.com/luarocks/lua-style-guide#modules). In short, create a local table that you then assign all the functions to, then return that table:

```LUA
local utilities = {}

function utilities.copy_prototype(...)

...

return utilities

File: data.lua [hunk 2]
```diff
@@ -0,0 +1,4 @@
+
+--- make data stage functions available globally
+---@usage flib.copy_prototype()
+flib = require("data.utilities")
```
Comment (raiguard) [inline data.lua]: As per the bit on modules, please don't add things to the global namespace. Modules should be required locally in the files where they are used:

`local utilities = require("__flib__.data.utilities")`

**Review by raiguard — CHANGES_REQUESTED** [review]

File: data/utilities.lua (hunk 1 above)
Comment (0ptera) [reply data/utilities.lua:5]: removed data.lua

File: data.lua (hunk 2 above)
Comment (0ptera) [reply data.lua]: moved functions into table

**Review by raiguard — APPROVED** [review]

**Pull request merged**